### PR TITLE
Refactor jukebox, jukebox datum, now jukebox audio is positional. And it fully respects deaf people. 

### PR DESCRIPTION
**Pull request by MrMelbert**
## About The Pull Request

- Jukebox is refactored into a datum that the rave visor and the jukebox uses. 

- Jukebox UI is now typescript. 

- How the Jukebox delivers sound to players has been rewritten. 
   - Now it adjusts the sound's position in accordance to where the listener is. 
      - This implementation was loosely inspired by that done by Baystation half a decade ago, so kudos to them.
   - Additionally, being deafened will temporarily mute the jukebox. 

- And sorry, in refactoring this I snuck in one tiny feature. 
   - You can now toggle looping on Jukeboxes to play the song foreeeverrrrr. 

## Why It's Good For The Game

It sounds wayyyyyy better. Overhead isn't even that bad, though it could be tested on a live server to make sure. 

https://github.com/tgstation/tgstation/assets/51863163/ec1321b6-bf1c-4c33-9663-83f2c23a4277

## Changelog

:cl: Melbert
refactor: Jukebox has been refactored. Jukebox music now updates as the player moves, mutes when the player is deafened, and overall sounds wayyy better. You can also now toggle song repeat on jukeboxes. 
/:cl:

### CODE REVIEW

File: code/game/machinery/dance_machine.dm [hunk 1]
```diff
@@ -153,60 +97,90 @@
...
 /obj/machinery/jukebox/disco/activate_music()
-	..()
+	. = ..()
+	if(!.)
+		return
 	dance_setup()
 	lights_spin()
+	begin_processing()
+
+/obj/machinery/jukebox/proc/stop_music()
```
Comment (san7890) [inline code/game/machinery/dance_machine.dm:160]: why is this in the middle of the code for the `disco` subtype? should be moved to be with the rest of the base-type code.

File: code/game/machinery/dance_machine.dm [hunk 2]
```diff
@@ -153,60 +97,90 @@
...
+
+/obj/machinery/jukebox/disco/stop_music()
+	. = ..()
+	if(!.)
+		return
+	QDEL_LIST(spotlights)
+	QDEL_LIST(sparkles)
+	end_processing()
+
+/obj/machinery/jukebox/on_set_is_operational(old_value)
```
Comment (san7890) [inline code/game/machinery/dance_machine.dm:181]: ditto, sandwiched in `disco` code

File: code/modules/mod/modules/modules_maint.dm [hunk 3]
```diff
@@ -80,48 +76,40 @@
...
-			continue
-		track.song_name = sound_params[1]
-		track.song_length = text2num(sound_params[2])
-		track.song_beat = text2num(sound_params[3])
-		songs[track.song_name] = track
-	if(length(songs))
-		var/song_name = pick(songs)
-		selection = songs[song_name]
+	music_player = new(src)
+	music_player.sound_loops = TRUE
```
Comment (san7890) [inline code/modules/mod/modules/modules_maint.dm:85]: Doesn't this need to be `QDEL_NULL`'d? (looking at the implementation you had for jukebox code)

**Review by san7890 — CHANGES_REQUESTED** [review]
one concern, rest are nits

File: code/datums/components/jukebox.dm [hunk 4]
```diff
@@ -0,0 +1,406 @@
...
+	// Because if we're muted for a different reason we don't wanna touch it
+	reason = ~reason
+
+	if((reason & MUTE_DEAF) && HAS_TRAIT(listener, TRAIT_DEAF))
+		return FALSE
+
+	if((reason & MUTE_PREF) && IS_PREF_MUTED(listener))
+		return FALSE
+
+	if((reason & MUTE_RANGE))
```
Comment (san7890) [inline code/datums/components/jukebox.dm:290]:
```suggestion
	if(reason & MUTE_RANGE)
```

File: code/game/machinery/dance_machine.dm [hunk 5]
```diff
@@ -153,60 +76,120 @@
...
+	/// Sparkle effects being played
+	VAR_PRIVATE/list/obj/effect/overlay/sparkles/sparkles = list()
+
+/obj/machinery/jukebox/disco/indestructible
+	name = "radiant dance machine mark V"
+	desc = "Now redesigned with data gathered from the extensive disco and plasma research."
+	req_access = null
+	anchored = TRUE
+	resistance_flags = INDESTRUCTIBLE | LAVA_PROOF | FIRE_PROOF | UNACIDABLE | ACID_PROOF
+	obj_flags = /obj::obj_flags | NO_DECONSTRUCTION
```
Comment (san7890) [inline code/game/machinery/dance_machine.dm:167]: i saw your other PR, should this be `parent_type`?

**Review by san7890 — CHANGES_REQUESTED** [review]
final nits i believe

File: code/game/machinery/dance_machine.dm (hunk 5 above)
Comment (MrMelbert) [reply code/game/machinery/dance_machine.dm:167]: Yes, it will conflict, I'll change it in this pr

**Comment by san7890** [comment]
conflicts

**Pull request merged**